### ISSUE TITLE
cookies again

### ISSUE THREAD

**Issue by Hugolarson** [JPro-one/JPro-tickets#94]
There is a issue with cookies I think.

WebAPI.getWebAPI(stage).setCookie("FOCUSED_STAGE", "stage1");
System.out.println(" 1 "+WebAPI.getWebAPI(stage).getCookies().get("FOCUSED_STAGE"));
console -> 1 stage1

WebAPI.getWebAPI(stage).setCookie("FOCUSED_STAGE", "stage2");
System.out.println(" 2 "+WebAPI.getWebAPI(stage).getCookies().get("FOCUSED_STAGE"));
console -> 2 stage1  (should be stage2)

Somehow the second time the cookie is set it does not register.

Tried by manipulating the map itself but exception is thrown
WebAPI.getWebAPI(stage).getCookies().put("FOCUSED_STAGE", "stage2") ; <- NPE

Also wonder what WebAPI.getWebAPI(stage).registerValue(string, String ) is used for? How can I read the registered value?

**Comment by FlorianKirmaier** [comment]
registerValue registers a value in the javascript environment. This can be used in combination with the executeScript, to do various things in the Browser.

The current logic is that JPro tells the browser to add a Cookie. When this is done, it informs JPro about the current cookies. This has a small delay, so putting cookies and reading them afterwards doesn't work instantaneously.

Have to think about whether we should change the design or keep it that way.

We will probably soon add something like "getSesssion()" "setSessionValue(key,value)" which doesn't go through the browser.

**Comment by Hugolarson** [comment]
Is session not slower than cookies since its stored in server?

**Comment by FlorianKirmaier** [comment]
Not in this case, because the JavaCode is running already in the server.

**Comment by Hugolarson** [comment]
Will "getSesssion()" be shared between tabs or will new tab create a new session?

**Comment by Hugolarson** [comment]
> registerValue registers a value in the javascript environment. This can be used in combination with the executeScript, to do various things in the Browser.
> 
> The current logic is that JPro tells the browser to add a Cookie. When this is done, it informs JPro about the current cookies. This has a small delay, so putting cookies and reading them afterwards doesn't work instantaneously.
> 
> Have to think about whether we should change the design or keep it that way.
> 
> We will probably soon add something like "getSesssion()" "setSessionValue(key,value)" which doesn't go through the browser.

Any progress on this new Session feature?

**Comment by FlorianKirmaier** [comment]
Hi @Hugolarson ,
thank you for reminding us.
not yet, but this feature is pretty high up in the priority list. 
Maybe about at the end of Juli, a SNAPSHOT build might be available. 
The variables would be shared between multiple tabs.

**Comment by Hugolarson** [comment]
Very good.

**Comment by f-geo** [comment]
Any new on this feature?

**Comment by Hugolarson** [comment]
Any news on this feature?

**Comment by FlorianKirmaier** [comment]
No, but good point. 
We are currently working on the next major update.
I put it into the Todo for this release.

**Comment by Hugolarson** [comment]
Hello,
It has been a while and I wonder about the status about the promise of Session feature/improvement.

BR, Hugo

**Comment by FlorianKirmaier** [comment]
Oh sorry, that was forgotten.
We will think next week about what to do about it - either adding it to the next release or providing a minimal library for it.

**Comment by Hugolarson** [comment]
Hi Floarian.
Any progress on this issue?

**Comment by FlorianKirmaier** [comment]
Oh, sorry for letting you wait so long.
Luckily, we now have time, and will do it immediately:

The change would be:

JPro Change:
 > Change Cookies to be autoupdated, when they are changed from somewhere else

As Library:
 > Provide a library for session management. Variables would only be accessible by the Server - and shared between multiple tabs.

**Comment by Hugolarson** [comment]
Very good!

**Comment by FlorianKirmaier** [comment]
@Hugolarson 
It's finally time!

#### JPro:
The latest SNAPSHOT build 2022.1.5-SNAPSHOT contains a new cookie implementation!
With the new implementation, any change is reflected in the WebAPI. Also when a cookie is changed/added/removed in another tab, or by another site/library it's updated in the JProServer.

#### Sessions:
It's implemented in a library.

You can find the implementation, of "variables only accessible in the server" here: https://github.com/JPro-one/jpro-utils
It provides an observable map, which is remembered when the server restarts. Its content is shared between multiple tabs, and its content is only accessible in the server and not in the browser.

Please provide feedback, if this is what you've expected, and whether it works for you!
Feel free to tell us, if there is something else, you could need.

**Comment by Hugolarson** [comment]
Hi,
I want to test but have encountered a problem and banging my head on the table since yesterday.
Have been a while using since using JPRO and started over with gradle sample you have on get started section.
I need to add a folder with dependency jars and read in forums that this should be put in buildscript / dependencies implementation fileTree(dir: "libs", include: ["*.jar"])
But it gives me error "Could not find method implementation() for arguments [directory 'libs'] on object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler."

**Comment by FlorianKirmaier** [comment]
Your Gradle version is probably very old.
Which version are you using?
They've changed the configuration "compile" to "implementation" some time ago, with a long deprecation phase.

**Comment by Hugolarson** [comment]
I'm using the wrapper in Hello-JPro.

Gradle 7.3.3
------------------------------------------------------------

Build time:   2021-12-22 12:37:54 UTC
Revision:     6f556c80f945dc54b50e0be633da6c62dbe8dc71

Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          11.0.16 (Azul Systems, Inc. 11.0.16+8-LTS)
OS:           Windows 10 10.0 amd64

**Comment by Hugolarson** [comment]
/**
 ******************  Script Configuration ******************
 */
buildscript {
  repositories {
    jcenter()
    gradlePluginPortal()

    maven {
      url "https://sandec.jfrog.io/artifactory/repo"
    }
  }

  dependencies {
  
    classpath 'com.sandec.jpro:jpro-plugin-gradle:2022.1.4'
    classpath 'org.openjfx:javafx-plugin:0.0.13'
	implementation fileTree(dir: "libs", include: ["*.jar"])
  }
}


/**
 ******************  Java Configuration ******************
 */
apply plugin: 'java-library'
apply plugin: 'application'
apply plugin: 'org.openjfx.javafxplugin'

compileJava {
  sourceCompatibility = 11
  targetCompatibility = 11
}

repositories {
  jcenter()
}

javafx {
    version = '17.0.1'
    modules = [ 'javafx.base', 'javafx.graphics', 'javafx.controls', 'javafx.fxml', 'javafx.media', 'javafx.web' ]
}



/**
 ******************  jpro Configuration ******************
 */
apply plugin: 'com.sandec.jpro'


/**
 * App Main Class
 */
//mainClassName = 'com.jpro.hellojpro.HelloJPro'
mainClassName = 'advit.vp_client.Client'

/**
 * jpro settings
 */
jpro {
  // for debugging
  // JVMArgs << '-agentlib:jdwp=transport=dt_socket,server=n,address=5006,suspend=y'

  JVMArgs << '-Xmx1000m'

  //jpro server port
  port = 8080

}

**Comment by FlorianKirmaier** [comment]
I think you'l have to define the dependencies, after applying the plugins.
It's imperative logic - the configurations only exist after the plugins are used.

**Comment by Hugolarson** [comment]
Got it. Thx.

**Comment by Hugolarson** [comment]
As I understand the JPro session is "stored" in the main Stage and some how in a cookie.
To get the session I need to have reference to the main stage. This is a bit difficult. 
Can't there be a way to get the session in a static way? Like WebApi.getSession?

**Comment by FlorianKirmaier** [comment]
Currently, JPro supports multiple users in one JVM, so using static usually doesn't work.
But you can configure the new JPro Loadbalancer, to do this. In this case, you can put the Stage (or webapi) into a static.

If you need more details/help on this topic, I would recommend you, send us an email, then we can have a short meeting, where we can give you details, and direct you into a direction most suited for you usecase.

**Comment by Hugolarson** [comment]
The application is a desktop application with global variables. I need some how to know "who I am" any where in code to access user data from a global map without passing a parameter around. It's a'lot of code to change and it's ugly.
Once I stored it in ThreadLocal but you told me that all users share same thread.

**Comment by FlorianKirmaier** [comment]
As I mentioned, you probably should use the JPro Loadbalancer https://www.jpro.one/docs/current/2.13/JPro_Loadbalancer_(Early_Access)
You can configure it, to have 1User per JVM, which might suite your usecasej.

**Comment by Hugolarson** [comment]
I read about the Load Balancer. One JVM per user seems to require a lot of server power.

**Comment by FlorianKirmaier** [comment]
Yes, but otherwise you can't use statics for user-specific data.
Otherwise, you will have to provide the "context of the user" when you want to do something user-specific

**Comment by Hugolarson** [comment]
It seems that now one can set and read session/cookie instantaneously. That is an improvement.

**Comment by Hugolarson** [comment]
Tried out sessionManager.getSession("key") and found out that its based upon a file created in user.home directory and does not need to have access to the Stage.

I think my problem can be solved if it instead of file on server, be based on cookie (or Local Storage) in browser. You think it's achievable for you to do this?

**Comment by FlorianKirmaier** [comment]
The cookie implementation of JPro already does this.
But you have to understand, that without context/stage - it's either 1 user per jvm, or it doesn't work properly.
So you will probably need to get the stage of the current user - whether you want it or not.
